### PR TITLE
[inductor]: Include convolution_backward in decide_layout_opt conv node detection

### PR DESCRIPTION
**Pull request by orrangetabby17**
# Motivation
When `AOTAutograd` splits a model into separate forward and backward graphs, the backward graph may contain only `convolution_backward` without a corresponding forward convolution.

Previously, `decide_layout_opt` 
https://github.com/pytorch/pytorch/blob/994f78b19267fdeb5292595db1fd4d5aa3d9b100/torch/_inductor/graph.py#L757-L779
only counted `aten.convolution.default` nodes when determining whether to enable layout optimization. As a result, backward-only graphs were incorrectly assigned `layout_opt=False`.
[PR#178945](https://github.com/pytorch/pytorch/pull/178945/changes#diff-3385fcbafbb6165d9e55060e0ddf92c46e17625eb6c0c38a07f32995950882da) add the `convolution_backward_lowering`
which is using the `layout_opt` (V.graph.layout_opt is False)to configure the layout of tensor.
https://github.com/pytorch/pytorch/blob/994f78b19267fdeb5292595db1fd4d5aa3d9b100/torch/_inductor/kernel/conv.py#L1088-L1098
As a result, the input tensor cant be set to NHWC layout. It caused Triton kernels to produce NCHW outputs in some cases (e.g., `timm`'s `mobilenetv3_large_100`).

# Main Change
In this PR we also count `aten.convolution_backward.default` as a conv node so backward graphs enable channels-last layout optimization. 

# Results
After this PR, `mobilenetv3_large_100` is **1.3× faster** on **Intel Data Center GPU** than the nightly PyTorch version.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @mcarilli @ptrblck @leslie-fang-intel @EikanWang @voznesenskym @penguinwu @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98 @xmfan @aditvenk

### CODE REVIEW

File: torch/_inductor/graph.py [hunk 1]
```diff
@@ -766,7 +766,10 @@ def decide_layout_opt(gm: GraphModule, *, is_inference: bool) -> bool:
             return True
 
         conv_nodes = [
-            n for n in gm.graph.nodes if n.target is torch.ops.aten.convolution.default
+            n for n in gm.graph.nodes if n.target in (
+                torch.ops.aten.convolution.default,
+                torch.ops.aten.convolution_backward.default,
```
Comment (guangyey) [inline torch/_inductor/graph.py:819]: I think this would introduce error for the other backend. See `return n.args[-1] > 1 and meta_val.size(1) > 1  # type: ignore[union-attr, operator]` defined in 814, for `torch.ops.aten.convolution_backward.default`, its last arg is `output_mask`, bool[3], can't be used in comparing with `meta_val.size(1)`.

Comment (orrangetabby17) [reply torch/_inductor/graph.py:819]: thanks, I will do some revise.

**Review by guangyey — COMMENTED** [review]
Now, this change will only apply to CPU/XPU. However, I think this should be benefit on all backends, including CUDA. Could you have a try whether CUDA can benefit from `convolution_backward`.

File: torch/_inductor/graph.py (hunk 1 above)
Comment (orrangetabby17) [reply torch/_inductor/graph.py:819]:
``` python
        conv_nodes = [
            n for n in gm.graph.nodes
            if n.target is torch.ops.aten.convolution.default
            or (
                n.target is torch.ops.aten.convolution_backward.default
                and any(n.args[idx].meta["val"].device.type in SUPPORTED_MKLDNN_DEVICES for idx in [0, 1])
            )
        ]
```
Changed to use the tensor device type to determine whether convolution_backward should be counted as a convolution node.

**Comment by orrangetabby17** [comment]
> Now, this change will only apply to CPU/XPU. However, I think this should be benefit on all backends, including CUDA. Could you have a try whether CUDA can benefit from `convolution_backward`.

I opened an issue (#189239) to discuss this problem in more detail. The current Inductor change is a workaround for XPU. We didn't implement the same workaround for CUDA because CUDA uses different heuristics to decide whether to enable channels-last layout optimization instead of using channels-last in every case.

File: torch/_inductor/graph.py [hunk 2]
```diff
@@ -766,7 +766,16 @@ def decide_layout_opt(gm: GraphModule, *, is_inference: bool) -> bool:
...
-            n for n in gm.graph.nodes if n.target is torch.ops.aten.convolution.default
+            n
+            for n in gm.graph.nodes
+            if n.target is torch.ops.aten.convolution.default
+            or (
+                n.target is torch.ops.aten.convolution_backward.default
+                and any(
+                    n.args[idx].meta["val"].device.type in SUPPORTED_MKLDNN_DEVICES
+                    for idx in [0, 1]
+                )
```
Comment (shunting314) [inline torch/_inductor/graph.py]: How does this work for cuda GPU? Ideally we should add the conv backward op disregarding the devices.

Comment (orrangetabby17) [reply torch/_inductor/graph.py]: This is an XPU-specific workaround, as oneDNN prefers the channels-last layout. CUDA relies on different heuristics to determine whether layout optimization should be enabled. Since the preferred layout optimization strategy differs across different accelerators, we cannot simply force the use of the channels-last layout.

**Comment by guangyey** [comment]
@orrangetabby17 Please take a look at the mps failure. I think they are related.

**Comment by orrangetabby17** [comment]
@claude is UT failure related to this PR.

File: torch/_inductor/graph.py (hunk 2 above)
Comment (shunting314) [reply torch/_inductor/graph.py]: Can you explain more why different kind of accelerators (Nvidia GPU v.s. XPU) have different layout preference for conv bwd?

Comment (orrangetabby17) [reply torch/_inductor/graph.py]: XPU always prefers the channels-last layout, so it bypasses decide_layout_opt() when running on XPU:

https://github.com/pytorch/pytorch/blob/994f78b19267fdeb5292595db1fd4d5aa3d9b100/torch/_inductor/graph.py#L782-L791

For CUDA, however, whether layout optimization is enabled is determined by a set of heuristics. The current heuristics are based on the presence of forward convolution nodes in the graph:

https://github.com/pytorch/pytorch/blob/994f78b19267fdeb5292595db1fd4d5aa3d9b100/torch/_inductor/graph.py#L810-L872

Supporting backward convolution would require extending these heuristics in a similar way. However, it's currently unclear what the expected CUDA behavior should be for backward-only graphs.

**Comment by orrangetabby17** [comment]
> @orrangetabby17 Please take a look at the mps failure. I think they are related.

Fixed.

**Review by guangyey — COMMENTED** [review]
This is not an XPU-specific issue. I suggest addressing it in a way that resolves the issue for all backends that may be affected.

**Review by eellison — COMMENTED** [review]
Thanks for fix - would you mind expanding a bit ? I looked into this with an agent on h100. It would be great if we could expand this to cuda as well. Here was codex's suggestion. Does this work for xpu as well?

• I checked this on CUDA. The saved forward tensors are already channels-last; the regression happens because the new backward lowering changes their layouts before falling back to
  ATen.

  Before #178945, NVIDIA used the direct ATen fallback with FX strides. I think we should restore that path when layout_opt=False, while making backward-only layout detection schema-
  aware.

  This changes MobileNetV3 training from 84.9 ms to 53.6 ms on H100. Five additional models ranged from neutral to 47% faster.
  <details>
  <summary>Proposed production diff</summary>

  diff --git a/torch/_inductor/graph.py b/torch/_inductor/graph.py
  @@
           conv_nodes = [
               n
               for n in gm.graph.nodes
               if n.target is torch.ops.aten.convolution.default
               or (
                   n.target is torch.ops.aten.convolution_backward.default
  -                and any(n.args[idx].meta["val"].device.type == "xpu" for idx in [0, 1])
  +                and (n.args[10][0] or n.args[10][1])
  +                and all(
  +                    n.args[idx].meta["val"].device.type in ("cuda", "xpu")
  +                    for idx in [0, 1, 2]
  +                )
               )
           ]

  +        def get_conv_tensor_args(n):
  +            if n.target is torch.ops.aten.convolution_backward.default:
  +                return n.args[:3]  # grad_output, input, weight
  +            return n.args[:2]      # input, weight
  +
  +        def get_conv_weight(n):
  +            return n.args[
  +                2 if n.target is torch.ops.aten.convolution_backward.default else 1
  +            ]
  +
  +        def get_conv_groups(n):
  +            return n.args[
  +                9 if n.target is torch.ops.aten.convolution_backward.default else -1
  +            ]
  +
           # Use these accessors in the existing dynamic-shape, grouped,
           # input/output-channel, and small-channel checks.

  diff --git a/torch/_inductor/kernel/conv.py b/torch/_inductor/kernel/conv.py
  @@
   def convolution_backward_lowering(...):
  +    device_type = ir.get_device_type(input)
  +    if device_type == "cuda" and not torch.version.hip and not V.graph.layout_opt:
  +        return aten_convolution_backward_fallback(
  +            grad_out,
  +            input,
  +            weight,
  +            bias_sizes,
  +            stride,
  +            padding,
  +            dilation,
  +            transposed,
  +            output_padding,
  +            groups,
  +            output_mask,
  +        )
  </details>

  Given this, I don’t think the current XPU-only version should land as-is.

**Comment by guangyey** [comment]
@orrangetabby17 Could you try the suggestion from eellison both on XPU and CUDA to see if it works as expected.

**Comment by orrangetabby17** [comment]
> Thanks for fix - would you mind expanding a bit ? I looked into this with an agent on h100. It would be great if we could expand this to cuda as well. Here was codex's suggestion. Does this work for xpu as well?
> 
> • I checked this on CUDA. The saved forward tensors are already channels-last; the regression happens because the new backward lowering changes their layouts before falling back to ATen.
> 
> Before #178945, NVIDIA used the direct ATen fallback with FX strides. I think we should restore that path when layout_opt=False, while making backward-only layout detection schema- aware.
> 
> This changes MobileNetV3 training from 84.9 ms to 53.6 ms on H100. Five additional models ranged from neutral to 47% faster.
> 
> Proposed production diff
> diff --git a/torch/_inductor/graph.py b/torch/_inductor/graph.py @@ conv_nodes = [ n for n in gm.graph.nodes if n.target is torch.ops.aten.convolution.default or ( n.target is torch.ops.aten.convolution_backward.default
> 
> * ```
>              and any(n.args[idx].meta["val"].device.type == "xpu" for idx in [0, 1])
>   ```
> 
> * ```
>              and (n.args[10][0] or n.args[10][1])
>   ```
> * ```
>              and all(
>   ```
> * ```
>                  n.args[idx].meta["val"].device.type in ("cuda", "xpu")
>   ```
> * ```
>                  for idx in [0, 1, 2]
>   ```
> * ```
>              )
>          )
>      ]
>   ```
> * ```
>      def get_conv_tensor_args(n):
>   ```
> * ```
>          if n.target is torch.ops.aten.convolution_backward.default:
>   ```
> * ```
>              return n.args[:3]  # grad_output, input, weight
>   ```
> * ```
>          return n.args[:2]      # input, weight
>   ```
> * ```
>      def get_conv_weight(n):
>   ```
> * ```
>          return n.args[
>   ```
> * ```
>              2 if n.target is torch.ops.aten.convolution_backward.default else 1
>   ```
> * ```
>          ]
>   ```
> * ```
>      def get_conv_groups(n):
>   ```
> * ```
>          return n.args[
>   ```
> * ```
>              9 if n.target is torch.ops.aten.convolution_backward.default else -1
>   ```
> * ```
>          ]
>   ```
> * ```
>      # Use these accessors in the existing dynamic-shape, grouped,
>      # input/output-channel, and small-channel checks.
>   ```
> 
> diff --git a/torch/_inductor/kernel/conv.py b/torch/_inductor/kernel/conv.py @@ def convolution_backward_lowering(...):
> 
> * device_type = ir.get_device_type(input)
> * if device_type == "cuda" and not torch.version.hip and not V.graph.layout_opt:
> * ```
>      return aten_convolution_backward_fallback(
>   ```
> * ```
>          grad_out,
>   ```
> * ```
>          input,
>   ```
> * ```
>          weight,
>   ```
> * ```
>          bias_sizes,
>   ```
> * ```
>          stride,
>   ```
> * ```
>          padding,
>   ```
> * ```
>          dilation,
>   ```
> * ```
>          transposed,
>   ```
> * ```
>          output_padding,
>   ```
> * ```
>          groups,
>   ```
> * ```
>          output_mask,
>   ```
> * ```
>      )
>   ```
> 
> Given this, I don’t think the current XPU-only version should land as-is.

Thanks for the detailed investigation. I'm currently looking into the CUDA side as well and checking what the right fix should be there. I'll also verify whether the same approach works cleanly for XPU.

**Comment by orrangetabby17** [comment]
> @orrangetabby17 Could you try the suggestion from eellison both on XPU and CUDA to see if it works as expected.
Sure, I'll test the suggested change on both XPU and CUDA and report back with the results.

**Comment by orrangetabby17** [comment]
@eellison 
Thanks so much for review. Two changes here:
1. **`decide_layout_opt` now also counts `convolution_backward` nodes on all devices.**
By counting convolution_backward as well, backward-only graphs can enable channels-last consistently with the forward graph, allowing this optimization to benefit all devices.

2. **Add an early-return gate in `convolution_backward_lowering` when the Triton path is disabled.**
On CUDA, the Triton convolution-backward kernels are force-disabled by #187081. The updated `conv.py` preserves CUDA's behavior from before this lowering was introduced.The `layout_opt && 2D` case still goes through the lowering, so CUDA can continue to use channels-last for the cuDNN backward fallback.

Performance:
On Intel Client GPUs, this change delivers a 1.3×+ speedup.
On an nvidia  4080 SUPER GPU, this change also delivers a 1.3×+ speedup.

**Comment by zou3519** [comment]
@claude review this please and put the review inside a GitHub collapsible section

**Review by guangyey — APPROVED** [review]

**Review by eellison — APPROVED** [review]
[Benchmark CI](https://github.com/pytorch/pytorch/actions/runs/35631539534/attempts/1) looks good. The ResNet18 slowdown didn’t reproduce on the [rerun](https://github.com/pytorch/pytorch/actions/runs/35631539534/job/106596637825), 
and CUDA-graph timing was essentially unchanged versus the [baseline](https://github.com/pytorch/pytorch/actions/runs/33577456871/job/106597405591).

  Can you fix the [Pyrefly error](https://github.com/pytorch/pytorch/actions/runs/33605201887/job/100259330944)? Then let’s land.

File: torch/_inductor/graph.py [hunk 3]
```diff
@@ -799,8 +799,25 @@ def decide_layout_opt(gm: GraphModule, *, is_inference: bool) -> bool:
...
+        # aten.convolution_backward prepends grad_output to aten.convolution's
+        # (input, weight) and takes groups at index 9 instead of last, so the
+        # heuristics below have to index the two schemas differently.
+        def is_conv_bwd(n: Node) -> bool:
+            return n.target is torch.ops.aten.convolution_backward.default
+
+        def get_conv_input_weight_vals(n: Node) -> tuple[Any, Any]:
+            if is_conv_bwd(n):
+                return n.args[1].meta["val"], n.args[2].meta["val"]
+            return n.args[0].meta["val"], n.args[1].meta["val"]
```
Comment (guangyey) [inline torch/_inductor/graph.py]:
```suggestion
            return n.args[0].meta["val"], n.args[1].meta["val"]  # type: ignore[union-attr, operator]
```

File: torch/_inductor/graph.py [hunk 4]
```diff
@@ -799,8 +799,25 @@ def decide_layout_opt(gm: GraphModule, *, is_inference: bool) -> bool:
...
 
+        # aten.convolution_backward prepends grad_output to aten.convolution's
+        # (input, weight) and takes groups at index 9 instead of last, so the
+        # heuristics below have to index the two schemas differently.
+        def is_conv_bwd(n: Node) -> bool:
+            return n.target is torch.ops.aten.convolution_backward.default
+
+        def get_conv_input_weight_vals(n: Node) -> tuple[Any, Any]:
+            if is_conv_bwd(n):
+                return n.args[1].meta["val"], n.args[2].meta["val"]
```
Comment (guangyey) [inline torch/_inductor/graph.py]:
```suggestion
                return n.args[1].meta["val"], n.args[2].meta["val"]  # type: ignore[union-attr, operator]
```

**Comment by guangyey** [comment]
Thanks @eellison 
@pytorchbot merge
